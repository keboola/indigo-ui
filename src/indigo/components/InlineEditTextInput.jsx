import React from 'react';
import PropTypes from 'prop-types';
import { Button, FormControl, OverlayTrigger, Tooltip } from 'react-bootstrap';
import Loader from './Loader';

class InlineEditTextInput extends React.Component {
  render() {
    if (this.props.isEditing) {
      return this._renderEditInput();
    }

    return this._renderStaticInput();
  }

  _onChange(e) {
    return this.props.onEditChange(e.target.value);
  }

  _onSubmit(e) {
    e.preventDefault();
    return this.props.onEditSubmit();
  }

  _renderEditInput() {
    return (
      <div className="form-inline kbc-inline-edit">
        <form onSubmit={this._onSubmit}>
          <FormControl
            type="text"
            value={this.props.text}
            disabled={this.props.isSaving}
            placeholder={this.props.placeholder}
            onChange={this._onChange}
          />

          <div className="kbc-inline-edit-buttons">
            <Button
              className="kbc-inline-edit-cancel"
              bsStyle="link"
              disabled={this.props.isSaving}
              onClick={this.props.onEditCancel}
            >
              <span className="kbc-icon-cross" />
            </Button>
            <Button
              className="kbc-inline-edit-submit"
              bsStyle="info"
              disabled={this.props.isSaving || !this.props.isValid}
              type="submit"
            >
              Save
            </Button>
            {this.props.isSaving && <Loader />}
          </div>
        </form>
      </div>
    );
  }

  _renderStaticInput() {
    const tooltip = (
      <Tooltip id="kbc-inline-edit-link-tooltip">
        {this.props.editTooltip}
      </Tooltip>
    );

    return (
      <OverlayTrigger placement={this.props.tooltipPlacement} overlay={tooltip}>
        <span className="kbc-inline-edit-link" onClick={this.props.onEditStart}>
          {this.props.text ? (
            <span>{this.props.text}</span>
          ) : (
            <span className="text-muted">{this.props.placeholder}</span>
          )}
          <span className="kbc-icon-pencil" />
        </span>
      </OverlayTrigger>
    );
  }
}

InlineEditTextInput.propTypes = {
  onEditStart: PropTypes.func.isRequired,
  onEditCancel: PropTypes.func.isRequired,
  onEditChange: PropTypes.func.isRequired,
  onEditSubmit: PropTypes.func.isRequired,
  text: PropTypes.string,
  isSaving: PropTypes.bool,
  isEditing: PropTypes.bool,
  isValid: PropTypes.bool,
  editTooltip: PropTypes.string,
  tooltipPlacement: PropTypes.string,
  placeholder: PropTypes.string
};

InlineEditTextInput.defaultProps = {
  placeholder: 'Click to edit',
  editTooltip: 'Click to edit',
  tooltipPlacement: 'top',
  isSaving: false,
  isEditing: false
};

export default InlineEditTextInput;
