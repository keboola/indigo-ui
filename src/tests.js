import renderer from 'react-test-renderer';

export const snapshot = (component) => {
  const tree = renderer
    .create(component)
    .toJSON();
  expect(tree).toMatchSnapshot();
};
