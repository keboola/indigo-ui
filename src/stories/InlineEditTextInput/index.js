import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import InlineEditTextInput from '../../indigo/components/InlineEditTextInput';

storiesOf('InlineEditTextInput', module)
  .add(
    'Read mode',
    withInfo({
      text: `
        Input in read only.
        `,
      inline: true
    })(() => {
      return (
        <InlineEditTextInput
          isEditing={false}
          isValid={true}
          onEditStart={() => null}
          onEditCancel={() => null}
          onEditChange={() => null}
          onEditSubmit={() => null}
          text="Original text"
          editTooltip="Click to edit"
          placeholder="Click to edit"
        />
      );
    })
  )
  .add(
    'Edit mode',
    withInfo({
      text: `
        Input in editing mode.
        `,
      inline: true
    })(() => {
      return (
        <InlineEditTextInput
          isEditing={true}
          isValid={true}
          onEditStart={() => null}
          onEditCancel={() => null}
          onEditChange={() => null}
          onEditSubmit={() => null}
          text="Original text"
          editTooltip="Click to edit"
          placeholder="Click to edit"
        />
      );
    })
  ).add(
    'Edit mode - saving',
    withInfo({
      text: `
        Input in editing mode.
        `,
      inline: true
    })(() => {
      return (
        <InlineEditTextInput
          isEditing={true}
          isValid={true}
          isSaving={true}
          onEditStart={() => null}
          onEditCancel={() => null}
          onEditChange={() => null}
          onEditSubmit={() => null}
          text="Original text"
          editTooltip="Click to edit"
          placeholder="Click to edit"
        />
      );
    })
  );
