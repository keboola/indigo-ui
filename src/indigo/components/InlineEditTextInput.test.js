import React from 'react';

import { snapshot } from '../../tests';
import InlineEditTextInput from './InlineEditTextInput';

describe('<InlineEditTextInput />', () => {
  const compulsoryProps = {
    onEditStart: () => null,
    onEditCancel: () => null,
    onEditChange: () => null,
    onEditSubmit: () => null
  };

  it('InlineEditTextInput - default', () => {
    snapshot(<InlineEditTextInput {...compulsoryProps} />);
  });

  it('InlineEditTextInput - read mode', () => {
    snapshot(<InlineEditTextInput {...compulsoryProps} isEditing={false} />);
  });

  it('InlineEditTextInput - edit mode', () => {
    snapshot(<InlineEditTextInput {...compulsoryProps} isEditing={true} />);
  });

  it('InlineEditTextInput - saving in edit mode', () => {
    snapshot(<InlineEditTextInput {...compulsoryProps} isEditing={true} isSaving={true} />);
  });

  it('InlineEditTextInput - disabled save button when text is not valid', () => {
    snapshot(<InlineEditTextInput {...compulsoryProps} isEditing={true} isValid={false} />);
  });
});
