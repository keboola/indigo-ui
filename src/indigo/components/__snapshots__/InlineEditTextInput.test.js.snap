// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineEditTextInput /> InlineEditTextInput - default 1`] = `
<span
  className="kbc-inline-edit-link"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <span
    className="text-muted"
  >
    Click to edit
  </span>
  <span
    className="kbc-icon-pencil"
  />
</span>
`;

exports[`<InlineEditTextInput /> InlineEditTextInput - disabled save button when text is not valid 1`] = `
<div
  className="form-inline kbc-inline-edit"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="form-control"
      disabled={false}
      id={undefined}
      onChange={[Function]}
      placeholder="Click to edit"
      type="text"
      value={undefined}
    />
    <div
      className="kbc-inline-edit-buttons"
    >
      <button
        className="kbc-inline-edit-cancel btn btn-link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="kbc-icon-cross"
        />
      </button>
      <button
        className="kbc-inline-edit-submit btn btn-info"
        disabled={true}
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;

exports[`<InlineEditTextInput /> InlineEditTextInput - edit mode 1`] = `
<div
  className="form-inline kbc-inline-edit"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="form-control"
      disabled={false}
      id={undefined}
      onChange={[Function]}
      placeholder="Click to edit"
      type="text"
      value={undefined}
    />
    <div
      className="kbc-inline-edit-buttons"
    >
      <button
        className="kbc-inline-edit-cancel btn btn-link"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="kbc-icon-cross"
        />
      </button>
      <button
        className="kbc-inline-edit-submit btn btn-info"
        disabled={true}
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;

exports[`<InlineEditTextInput /> InlineEditTextInput - read mode 1`] = `
<span
  className="kbc-inline-edit-link"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <span
    className="text-muted"
  >
    Click to edit
  </span>
  <span
    className="kbc-icon-pencil"
  />
</span>
`;

exports[`<InlineEditTextInput /> InlineEditTextInput - saving in edit mode 1`] = `
<div
  className="form-inline kbc-inline-edit"
>
  <form
    onSubmit={[Function]}
  >
    <input
      className="form-control"
      disabled={true}
      id={undefined}
      onChange={[Function]}
      placeholder="Click to edit"
      type="text"
      value={undefined}
    />
    <div
      className="kbc-inline-edit-buttons"
    >
      <button
        className="kbc-inline-edit-cancel btn btn-link"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        <span
          className="kbc-icon-cross"
        />
      </button>
      <button
        className="kbc-inline-edit-submit btn btn-info"
        disabled={true}
        type="submit"
      >
        Save
      </button>
      <span
        className="fa fa-spin fa-spinner"
      />
    </div>
  </form>
</div>
`;
